### PR TITLE
fix: limitless plugin - reduce extraneous conns made in scenarios whe…

### PR DESCRIPTION
**Pull request by aaronchung-bitquill**
…re router cache is empty

### By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java [hunk 1]
```diff
@@ -177,10 +172,7 @@ private Connection connectInternalWithDialect(
           new Object[] {selectedHostSpec != null ? selectedHostSpec.getHost() : "null"}));
     } catch (SQLException e) {
       LOGGER.warning(Messages.get("LimitlessConnectionPlugin.errorSelectingRouter", new Object[] {e.getMessage()}));
-      if (conn == null  || conn.isClosed()) {
-        conn = connectFunc.call();
-      }
-      return retryConnectWithLeastLoadedRouters(limitlessRouters, props, conn, hostSpec);
+      return retryConnectWithLeastLoadedRouters(limitlessRouters, props, connectFunc.call(), connectFunc, hostSpec);
```
Comment (aaronchung-bitquill) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java:175]: wondering in these cases where there is no connection object already instantiated, if i should just leave the connection param to be null.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java [hunk 2]
```diff
@@ -142,29 +142,24 @@ private Connection connectInternalWithDialect(
...
-        limitlessRouters = synchronouslyGetLimitlessRoutersWithRetry(conn, hostSpec.getPort(), props);
+        limitlessRouters = synchronouslyGetLimitlessRoutersWithRetry(null, connectFunc, hostSpec.getPort(), props);
       } else {
         LOGGER.finest(Messages.get("LimitlessConnectionPlugin.usingProvidedConnectUrl"));
-        return conn;
+        return connectFunc.call();
       }
     }
 
     if (limitlessRouters.contains(hostSpec)) {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java:156]: Let's check for null here

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java [hunk 3]
```diff
@@ -177,10 +172,7 @@ private Connection connectInternalWithDialect(
...
     } catch (SQLException e) {
       LOGGER.warning(Messages.get("LimitlessConnectionPlugin.errorSelectingRouter", new Object[] {e.getMessage()}));
-      if (conn == null  || conn.isClosed()) {
-        conn = connectFunc.call();
-      }
-      return retryConnectWithLeastLoadedRouters(limitlessRouters, props, conn, hostSpec);
+      return retryConnectWithLeastLoadedRouters(limitlessRouters, props, connectFunc.call(), connectFunc, hostSpec);
     }
 
     try {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java:178]: selectedHostSpec can be null

File: wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java [hunk 4]
```diff
@@ -252,8 +242,8 @@ private Connection retryConnectWithLeastLoadedRouters(
 
     while (retryCount++ < maxRetries) {
       if (currentRouters.stream().noneMatch(h -> h.getAvailability().equals(HostAvailability.AVAILABLE))) {
-        if (conn != null && !conn.isClosed()) {
-          currentRouters = synchronouslyGetLimitlessRoutersWithRetry(conn, hostSpec.getPort(), props);
+        if ((conn != null && !conn.isClosed())) {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java:245]: What is conn is null but there's a connectFunc?

**Comment by aaronchung-bitquill** [comment]
starting again on a new branch

**Pull request closed without merging**